CXF-8442 Java 16 & CachedOutputStream Fix IOException closed after multiple close() & postClose() calls

Sevyls opened the pull request:
see https://issues.apache.org/jira/browse/CXF-8442

reta commented on core/src/main/java/org/apache/cxf/io/CachedOutputStream.java:
```diff
@@ -225,7 +225,11 @@ public void close() throws IOException {
         }
         if (!maybeDeleteTempFile(currentStream)) {
-            postClose();
+            try {
```
@Sevyls thank you for the investigation and the consequent suggested fix. I think we should attack the problem from the different angle (as you also mentioned in the ticket):
 - `postClose()` could raise `IOException` so does `close()` and this is legitimate, no need to catch it
 - however, as you found out, `CacheAndWriteOutputStream` may attempt to close the underlying stream many times
I think the proper solution would be:
 - `CacheAndWriteOutputStream`  should keep track if the `flowThroughStream` has been closed already 
 - `closeFlowthroughStream()` and `postClose()` should set or/and check if the `flowThroughStream` has been closed already and behave appropriately (flush/close or do nothing)

What do you think?

Sevyls commented:
Hi @reta,

Thanks for your answer and suggestions. I reverted my changes in `CacheAndWriteOutputStream`.

I added a variable isClosed to `CacheAndWriteOutputStream` and as `postClose()` and `closeFlowthroughStream()` had the same implementation I reduced it to a call to `postClose()`. 

Please review my pull request again, see the test case `CacheAndWriteOutputStreamTest#testCloseMultipleTimes()`.

reta commented on core/src/main/java/org/apache/cxf/io/CacheAndWriteOutputStream.java:
```diff
@@ -47,19 +47,32 @@ public void setCacheLimit(long l) {
+    @Override
+    protected void doClose() throws IOException {
+        super.doClose();
+        isClosed = true;
```
Just for the record, probably not necessary (since `close()` would call `postClose()`) but won't hurt

reta replied:
@Sevyls I had to remove `close()` and `doClose()`, the complex relationships between these methods led to the issues that `isClosed` flag being set **before** closing the streams. As the results, streams stayed open.

reta commented on core/src/main/java/org/apache/cxf/io/CacheAndWriteOutputStream.java:
```diff
@@ -29,7 +29,7 @@
  */
 public class CacheAndWriteOutputStream extends CachedOutputStream {
-
+    private boolean isClosed = false;
```
The stream should not be shared between different threads, so we may not need `volatile` or atomic

reta commented:
@Sevyls Thank you.
